d2/src/main/java/com/linkedin/d2/balancer/clients/DegraderTrackerClient.java

```diff
@@ -0,0 +1,187 @@
+    {
+      config = new DegraderImpl.Config();
+    }
+
+    config.setCallTracker(_callTracker);
+    config.setClock(clock);
+    // The overrideDropRate will be globally determined by the DegraderLoadBalancerStrategy.
+    config.setOverrideDropRate(0.0);
+
+    if (uriSpecificProperties == null)
+    {
+      uriSpecificProperties = new HashMap<>();
+    }
+    if (uriSpecificProperties.containsKey(PropertyKeys.DO_NOT_SLOW_START)
+      && Boolean.parseBoolean(uriSpecificProperties.get(PropertyKeys.DO_NOT_SLOW_START).toString()))
+    {
+      config.setInitialDropRate(DegraderImpl.DEFAULT_DO_NOT_SLOW_START_INITIAL_DROP_RATE);
+    }
+
+    /* TrackerClient contains state for each partition, but they actually share the same DegraderImpl
+     *
+     * There used to be a deadlock if each partition has its own DegraderImpl:
+     * getStats() and rolloverStats() in DegraderImpl are both synchronized. getstats() will check whether
+     * the state is stale, and if yes a rollover event will be delivered which will call rolloverStats() in all
+     * DegraderImpl within this CallTracker. Therefore, when multiple threads are calling getStats() simultaneously,
+     * one thread may try to grab a lock which is already acquired by another.
+     *
+     * An example:
+     * Suppose we have two threads, and here is the execution sequence:
+     * 1. Thread 1 (DegraderImpl 1): grab its lock, enter getStats()
+     * 2. Thread 2 (DegraderImpl 2): grab its lock, enter getStats()
+     * 3. Thread 1: PendingEvent is delivered to all registered StatsRolloverEventListener, so it will call rolloverStats()
+     *    in both DegraderImpl 1 and DegraderImpl 2. But the lock of DegraderImpl 2 has already been acquired by thread 2
+     * 4. Same happens for thread 2. Deadlock.
+     *
+     * Solution:
+     * Currently all DegraderImpl within the same CallTracker actually share exactly the same information,
+     * so we just use create one instance of DegraderImpl, and use it for all partitions.
+     *
+     * Pros and Cons:
+     * Deadlocks will be gone since there will be only one DegraderImpl.
+     * However, now it becomes harder to have different configurations for different partitions.
+     */
+    int mapSize = partitionDataMap.size();
+    Map<Integer, PartitionState>partitionStates = new HashMap<Integer, PartitionState>(mapSize * 2);
+    config.setName("TrackerClient Degrader: " + uri);
+    DegraderImpl degrader = new DegraderImpl(config);
+    DegraderControl degraderControl = new DegraderControl(degrader);
+    for (Map.Entry<Integer, PartitionData> entry : partitionDataMap.entrySet())
+    {
+      int partitionId = entry.getKey();
+      PartitionState partitionState = new PartitionState(entry.getValue(), degrader, degraderControl);
+      partitionStates.put(partitionId, partitionState);
+    }
+    _partitionStates = Collections.unmodifiableMap(partitionStates);
+  }
+
+  public Degrader getDegrader(int partitionId)
```

Comment:
I'm just keeping existing degrader logic as similar as it was to before with minimal changes.  Degrader strategy will use DegraderTrackerClient impl while any other strategies going forward will use TrackerClient interface.

Reply:
I'd rather expose TrackerClient interface and hide the implementation behind.

Reply:
To add a few more points: DegraderTrackerClient is different from TrackerClientImpl because it is supposed to only serve DEGRADER strategy. To serve DEGRADER strategy, we have additional public methods in DegraderTrackerClient, such as getDegrader, we do not want to put this into the interface because the other strategy would never need this.